### ISSUE TITLE
[Thinking out loud] - Dynamic/interpolated/concatenated classes?

### ISSUE THREAD

**Issue by VicGUTT** [tailwindlabs/tailwindcss#3941]
I'm pumped by this new feature! Awesome work!
The on demand feature and arbitrary values support got me excited for a possible way to allow concatenated classes with Tailwind.

We all know currently with Tailwind **it is important to avoid dynamically creating class strings in our templates with string concatenation, otherwise PurgeCSS won't know to preserve those classes.** _[Docs](https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html)_

And on the other hand, even with the new JIT mode, classes like `bg-$myVariable` _(PHP)_ and `bg-${myVariable}` _(JS)_ are evaluated at runtime and Tailwind will never be able to know what the resulting class will be.

That said, we could create a new syntax: `prefix-{token}` or `prefix-{*}`, that would instruct Tailwind to generate all possible values for that _prefix + { + token + }_ combo, example:
```html
<button
    class="
        mt-{spacing} or mt-{*}
        text-{colors} or text-{*}
        bg-{colors}-500 or bg-{*}-500

        mt-$foo
        text-$bar
        bg-$baz-500
    "
>

<!--
    Would generate:
        - .mt-px .mt-0 .mt-0.5 .mt-1 ... .mt-96
        - text-white .text-indigo-50 ... .text-indigo-900 ... .text-blue-50 ... .text-blue-900
        - .bg-indigo-500 .bg-pink-500 ... .bg-gray-500
-->
```

But I'm not convinced by this idea, and it might even be out of scope for the JIT feature.

On the other hand, and in the spirit of the on demand classes with arbitrary values like `w-[calc(100%+1rem)]` I think we could introduce CSS custom properties: `w-[var(--myVariable,fallback)]` or better, **Tailwind custom properties**: `w-[--tw-w-4]` which would allow us to get as close as possible to dynamic classes.

Usage example:

```html
<button class="bg-[var(--myVariable,2rem)] w-[--tw-w-4]">

<!--
    Would generate:
        - .bg-\[var\(--myVariable\,2rem\)\] {
            background-color: var(--myVariable, 2rem);
        }
        - .w-\[--tw-w-4\] {
            width: 1rem;
            or
            width: var(--tw-w-4, 1rem);
        }
-->
```

Like said in the issue title, I am currently just thinking out loud. And with the features this new JIT brings, being able to handle dynamically constructed classes is probably the last remaining _annoying_ thing with/limitation of Tailwind.

This issue is meant as a "food for thought" idea and would be better suited for Github's Discussions feature if it was available for this repository.

**Comment by benjaminpreiss** [comment]
@VicGUTT I love the idea! Nevertheless, I think it would be better to have something like runcss as a runtime complement. That way we can keep the syntax?

**Comment by VicGUTT** [comment]
I did not know of RunCSS but the idea is indeed interesting and I love out of the box thinking like this. I don't think I would use a similar approach in production though as I would not want my users to pay any byte/latency/CPU etc... penalty. And if I were, I would most likely use the ["Tailwindcss JIT CDN"](https://github.com/beyondcode/tailwindcss-jit-cdn)

**Comment by benjaminpreiss** [comment]
Yeah, @VicGUTT I feel you... I didn't know about the CDN, so that might be a valid alternative. Nevertheless, I think a little extra CPU is a small price for keeping the syntax uniform. What happens if tailwind changes the underlying CSS for a class? Then we have to go search in the style tags for the custom CSS we added to substitute a tailwind class

**Comment by VicGUTT** [comment]
I'm not sure I'm following. What do you mean by "_keeping the syntax uniform_" ? If Tailwind ever changes the underlying rules referenced by a class that would probably mean we made a conscious choice to jump a major version _(from v2.x.x to v3.0.0 etc...)_

**Comment by benjaminpreiss** [comment]
@VicGUTT True!
